lnbits/extensions/boltz/boltz.py

```diff
@@ -0,0 +1,458 @@
+            extra={"tag": "boltz", "swap_id": swap_id},
+        )
+    except Exception as e:
+        raise HTTPException(status_code=HTTPStatus.INTERNAL_SERVER_ERROR, detail=str(e))
+
+    refund_privkey = ec.PrivateKey(os.urandom(32), True, net)
+    refund_pubkey_hex = hexlify(refund_privkey.sec()).decode("UTF-8")
+
+    res = create_post_request(
+        BOLTZ_URL + "/createswap",
+        {
+            "type": "submarine",
+            "pairId": "BTC/BTC",
+            "orderSide": "sell",
+            "refundPublicKey": refund_pubkey_hex,
+            "invoice": payment_request,
+        },
+    )
+    logger.info(f" - created swap, boltz_id: {res['id']}. wallet: {data.wallet}")
+    return SubmarineSwap(
+        id=swap_id,
+        time=getTimestamp(),
+        wallet=data.wallet,
+        amount=data.amount,
+        refund_privkey=refund_privkey.wif(net),
+        refund_address=data.refund_address,
+        boltz_id=res["id"],
+        status="pending",
+        address=res["address"],
+        expected_amount=res["expectedAmount"],
+        timeout_block_height=res["timeoutBlockHeight"],
+        bip21=res["bip21"],
+        redeem_script=res["redeemScript"],
+    )
+
+
+def get_fee_estimation() -> int:
+    # hardcoded maximum tx size, in the future we try to get the size of the tx via embit (not possible yet)
+    tx_size_vbyte = 200
+    mempool_fees = get_mempool_fees()
+    return mempool_fees * tx_size_vbyte
+
+
+# claim tx for reverse swaps
+# refund tx for normal swaps
+async def create_onchain_tx(
+    swap: Union[ReverseSubmarineSwap, SubmarineSwap], mempool_lockup_tx
+) -> Transaction:
+
+    if type(swap) == SubmarineSwap:
+        current_block_height = get_mempool_blockheight()
+        if current_block_height <= swap.timeout_block_height:
+            msg = f"refund not possible, timeout_block_height ({swap.timeout_block_height}) is not yet exceeded ({current_block_height})"
+            raise HTTPException(status_code=HTTPStatus.METHOD_NOT_ALLOWED, detail=msg)
+        privkey = ec.PrivateKey.from_wif(swap.refund_privkey)
+        preimage = b""
+        onchain_address = swap.refund_address
+        sequence = 0xFFFFFFFE
+    else:
+        privkey = ec.PrivateKey.from_wif(swap.claim_privkey)
+        preimage = unhexlify(swap.preimage)
+        onchain_address = swap.onchain_address
+        sequence = 0xFFFFFFFF
+
+    locktime = swap.timeout_block_height
+    redeem_script = unhexlify(swap.redeem_script)
+
+    fees = get_fee_estimation()
+
+    tx, txid, vout_cnt, vout_amount = mempool_lockup_tx
+
+    script_pubkey = script.address_to_scriptpubkey(onchain_address)
+
+    vin = [TransactionInput(unhexlify(txid), vout_cnt, sequence=sequence)]
+    vout = [TransactionOutput(vout_amount - fees, script_pubkey)]
+    tx = Transaction(vin=vin, vout=vout)
+
+    if type(swap) == SubmarineSwap:
+        tx.locktime = locktime
+
+    # TODO: 2 rounds for fee calculation, look at vbytes after signing and do another TX
```

Comment:
Todo: still pending?

Reply:
yep, that todo is still pending, because embit has no ability to return the vbyte size of the TX yet.